Experimental: Algebraic Statistics

bkholler opened the pull request:
This is work in progress for implementing graphical and phylogenetic models from algebraic statistics. This builds upon the draft pull request "Adding Graphical Models in Experimental" #3621.

Commit:
Add fist files graphical models

bkholler committed:
added folders for algebraic statistics

mariusneu committed:
test commit

chgoergen committed:
added a greeting to readme

chgoergen committed:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

chgoergen committed:
TEST added a greeting to the phylomodels

mariusneu committed:
reverted test commit

mariusneu committed:
Merge remote-tracking branch 'origin/HEAD'

chgoergen committed:
TEST from home

chgoergen committed:
added phylomodels containing a few models, untested

mariusneu committed:
first edits on specft

mariusneu committed:
finished specialized fourier transform and its inverse

mariusneu committed:
corrected a computational mistake in inverse spec_ft

Commit:
Fourier Parametrization

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

mariusneu committed:
deleted duplicate functionality

mariusneu committed:
Merge branch 'spec_ft' into master  ->
Marina's Code and my Code combined.

mariusneu committed:
changed function names, minor changes in comments

mariusneu committed:
Split the code into totally three files according to the old structure we agreed upon.

mariusneu committed:
Deleted Christiane's phylomodels.jl as I integrated the contents into PhylogeneticModels.jl

mariusneu committed:
Edited Christiane's constructors to work with the new PhylogeneticModel struct

mariusneu committed:
Added include-statements

mariusneu committed:
corrected minor mistake in which_group_element, added correct include statements in GraphicalModels.jl

mariusneu committed:
corrected minor spelling mistake

mariusneu committed:
corrected two minor mistakes in the building of the model

Commit:
Fixed typos K2 and K3 parameters

bkholler committed:
added data structures for gaussians and graphical models

chgoergen committed:
added raw docs

Commit:
cherries function

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

chgoergen committed:
changed order in struct

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

Commit:
adding documentation

Commit:
added runtests

marinagarrote committed:
Update README.md

Commit:
export in GraphicalModels.jl

taboege committed:
Fix compilation

taboege committed:
Add MarkovRing and CI statements

taboege committed:
Supply equality and hash methods for CIStmt

bkholler committed:
added parameterization of directed gaussian graphical models

bkholler committed:
Merge branch 'bgh/AlgebraicStatistics' of github.com:bkholler/Oscar.jl into bgh/AlgebraicStatistics

taboege committed:
Fix keyword argument syntax

bkholler committed:
added parameterizations and vanishing ideals for directed and undirected gaussian graphical models

bkholler committed:
Merge branch 'bgh/AlgebraicStatistics' of github.com:bkholler/Oscar.jl into bgh/AlgebraicStatistics

chgoergen committed:
started playing around with documentation as md document

Commit:
Added GroupBasedPhylogeneticModel and fixed the corresponding methods

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

Commit:
Fixed equivalent classes functions

Commit:
Using hadamard(R::ZZMatrixSpace) method

Commit:
simplified code for specialized_fourier_transform and inverse

Commit:
update test files and exports

chgoergen committed:
added all exported functions to .md doc

chgoergen committed:
added raw doc to all models

chgoergen committed:
started documenting parametrisation, nothing substantial yet

chgoergen committed:
added possible documentation for parametrisations

Commit:
commit to pull

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

chgoergen committed:
added doc for equivalent classes

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

Commit:
changed access keys in parametrization map

Commit:
changed entries transition matrices in GMM

Commit:
added root distribution in parametrization

Commit:
Text to print in PhylogeneticModels and GroupBasedPM

chgoergen committed:
completed the docs for all models and exported functions, modified .md for clearer visualisation, changed (and broke) function (inverse)specialised_fourier_transform by calling pm only -- docs are done now

Commit:
fixed equivalent classes

Commit:
fixed sum_equivalent_classes

Commit:
recent version with comments for everyone to work on

chgoergen committed:
fixed indentation, blank lines, comments, and imperative form in docs

chgoergen committed:
fixed indentation in docs

chgoergen committed:
renamed file, fixed indentation in doc, put questions in slack

chgoergen committed:
changed parametrisation to parametrization

Commit:
updated testfile

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

Commit:
updated todos

Commit:
test for models including affine complete

Commit:
Merge branch 'master' of github.com:marinagarrote/Oscar.jl

Commit:
tests for parametrizations

Commit:
fixed tests and documentation

Commit:
Fixing documentation PhylogeneticModels.jl

Commit:
Adding sections/comments

Commit:
Fixing documentation PhyloParametrization.jl

Commit:
changes order_eges for sort_edges

Commit:
fixing doctests

Commit:
fixing doctests

Commit:
changed spaces and lenght output

Commit:
fixing problems documentation

Commit:
ordering output trans_mat and fourier_param

Commit:
ordering output trans_mat and fourier_param 2

Commit:
fix output groupBasedModel struct

bkholler committed:
added documentation for gaussian graphical models and some functionality for undirected discrete graphical models

taboege committed:
AlgebraicStatistics: document CI.jl and Markov.jl

Commit:
Renaming folder to AlgebraicStatistics

bkholler committed:
added documentation pages for graphical models and gaussian graphical models

bkholler committed:
Merge branch 'bgh/AlgebraicStatistics' of github.com:bkholler/Oscar.jl into bgh/AlgebraicStatistics

bkholler committed:
added intro documentation

taboege committed:
Merge remote-tracking branch 'marina/bggn/phylogenetics' into bgh/AlgebraicStatistics

taboege committed:
Merge remote-tracking branch 'origin/bgh/AlgebraicStatistics' into bgh/AlgebraicStatistics

fingolfin commented:
@bkholler thanks for you contribution! Are you collaborating with @marinagarrote? I am wondering because that PR is still marked as draft, yours is not, so what is the plan moving forward here -- shall this PR here supersede the other?

fingolfin commented on docs/oscar_references.bib:
```diff
@@ -672,6 +672,22 @@ @Article{Der99
 }
 
+@Book{DSS09,
+ Author = {Drton, Mathias and Sturmfels, Bernd and Sullivant, Seth},
```
Please take a look at https://docs.oscar-system.org/dev/DeveloperDocumentation/documentation/#Updating-the-bibliography for how to update the bibliography (if you have trouble getting `bibtool` to work, someone can also help you and update this in the PR for you, just let us know)

taboege replied:
Done, the test passes now.

fingolfin commented on experimental/AlgebraicStatistics/README.md:
```diff
@@ -0,0 +1,13 @@
+## Status
+
+We plan to also provide a function to automatically copy this template 
+for you to start your own package. 
```
Clearly this is not yet ready ;-)

taboege replied:
Added a README with our aims and some information about the status.

bkholler commented:
@fingolfin Yes! @marinagarrote originally started developing some graphical models and phylogenetics functionality separately but are now working together as team to create one algebraic statistics package. We just merged her fork into mine and this PR should now supersede that one. Sorry for the confusion.

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+        new_Q = [p for p in Q if p in neighbors(G, v)]
+        new_Y = [p for p in Y if p in neighbors(G, v)]
+
+        cliques = vcat(cliques, bron_kerbosch(G, new_R, new_Q, new_Y))
```
```suggestion
        append!(cliques, bron_kerbosch(G, new_R, new_Q, new_Y))
```

bkholler replied:
I've removed this file for now since it is still far from ready and wasn't supposed to be in this PR but we'll definitely make these changes when we add it back in later.

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+
+    for v in Q
+
+        new_R = vcat(R, v);
```
No need for those semicolons :-)
```suggestion
        new_R = vcat(R, v)
```

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+    cliques = []
+
+    if length(P) == 0 && length(X) == 0
+        
```
```suggestion
```

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+function bron_kerbosch(G, R, P, X)
+
+    cliques = []
```
Note that this will be a `Vector{Any}`, leading to type unstable code down the line. If you have any information about what the content might be, you could let Julia know: If `T` is a supertype of the content, you can use
```suggestion
    cliques = T[]
```
to get an empty `Vector{T}` here.

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+@doc raw"""
+    ring(R::GaussianRing)
+
+Return the Oscar multivariate polynomial ring inside the `GaussianRing`.
```
```suggestion
Return the multivariate polynomial ring inside the `GaussianRing`.
```
or even better
```suggestion
Return the multivariate polynomial ring inside `R`.
```

taboege replied:
Done.

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+    graphical_model(G::Graph{Undirected}, S::GaussianRing, k_var_name::String="k")
+
+A parametric statistical model associated to an undirected graph.
+It contains an undirected graph `G`, a GaussianRing `S` where the vanishing ideal of the model naturally lives, 
```
If you mean "GaussianRing" as in "the type name", then it should be formatted as code, i.e.:
```suggestion
It contains an undirected graph `G`, a `GaussianRing` `S` where the vanishing ideal of the model naturally lives, 
```
But perhaps you rather mean this mathematical (makes more sense to me, esp. since the type is already specified in the first line of the docstring, as part of the method signature), I suggest writing here and elsewhere:
```suggestion
It contains an undirected graph `G`, a Gaussian ring `S` where the vanishing ideal of the model naturally lives, 
```

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+
+
+
+maximal_cliques(G)
```
This code is on the top level of the file?

bkholler replied:
Ah sorry I did not mean for file "DiscreteGraphicalModels.jl" to be included in the pull request at all actually hence all the weird errors and lack of comments or documentation. I just started working on that and it is far from ready but we wanted to submit a first pull request so we could get some early feedback on what we have currently.

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
````diff
@@ -0,0 +1,86 @@
+(1, 2), (2, 3)
+```
+"""
+function graphical_model(G::Graph{Undirected}, S::MarkovRing, t_var_name::String="t")
````
Here and elsewhere where `BLA_var_name` occurs: I recommend replacing using our `VarName` type, which makes it more flexible (also symbols and chars can be passed in) while also making it possible drop the `_var_name` suffix which is rather unusual:
```suggestion
function graphical_model(G::Graph{Undirected}, S::MarkovRing, t::VarName="t")
```
or
```suggestion
function graphical_model(G::Graph{Undirected}, S::MarkovRing, t::VarName=:t)
```

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+"""
+function graphical_model(G::Graph{Undirected}, S::MarkovRing, t_var_name::String="t")
+
+    cliques = [reduce(*, string.(C)) for C in maximal_cliques(G)]
```
Isn't this the same as
```suggestion
    cliques = [join(C) for C in maximal_cliques(G)]
```
or perhaps even
```suggestion
    cliques = join.(maximal_cliques(G))
```

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+
+    rvs = random_variables(R)
+
+    R, t = polynomial_ring(QQ, vcat([t_var_name*C*string([i]) for C in cliques for i in ])
```
`for i in ]` -> something is missing after the `in` ?

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+function bron_kerbosch(G, R, P, X)
```
Can you restrict the argument types at all? Not so much for the compiler (it doesn't care) but for people trying to read the code it would help if one knew what kind of object G, R, P, X possibly could be :-)

fingolfin commented on experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl:
```diff
@@ -0,0 +1,86 @@
+
+        new_R = vcat(R, v);
+        new_Q = [p for p in Q if p in neighbors(G, v)]
+        new_Y = [p for p in Y if p in neighbors(G, v)]
```
Depending on what types `Q` and `neighbors(G,v)` are, you might also be able to write
```suggestion
        nb = neighbors(G, v)
        new_Q = intersect(Q, nb)
        new_Y = intersect(Q, nb)
```

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+struct GaussianRing
+    ring::Ring
+    gens::Dict
+    covariance_matrix
```
Can you specify at least a supertype? Like, is this a Julia `Matrix`, or an Oscar `MatElem`, or something else?

taboege replied:
Done. It should always be `MatElem`.

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+    coeffs = base_ring(R.ring)
+    k = ngens(R.ring)
+   
+    print("Gaussian ring over $(coeffs) in $(k) variables", "\n", chop(string(gens(R.ring)), head = 16, tail = 1))
```
```suggestion
    print(io, "Gaussian ring over $(coeffs) in $(k) variables", "\n", chop(string(gens(R.ring)), head = 16, tail = 1))
```

taboege replied:
Done everywhere.

fingolfin commented:
@bkholler that's totally fine! thank you for the explanation, now it is clearer to me what is going on. If you don't need the old PR anymore, feel free to close it.

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+
+struct GaussianRing
+    ring::Ring
+    gens::Dict
```
Based on what you wrote below, maybe the type here can be strengthened?
```suggestion
    gens::Dict{Tuple{Int64, Int64}, <:PolyRingElem}
```

taboege replied:
I just committed a number of strengthenings to the types of `MarkovRing` and `GaussianRing`.

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+
+
+struct GaussianRing
+    ring::Ring
```
In your constructor, it actually seems to be limited to fields?

taboege replied:
Fields are the most reasonable assumption. I have used that throughout.

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+
+    S, s = polynomial_ring(base_ring, reduce(vcat, [[s_var_name*string([i,j]) for j in i:n] for i in 1:n]))
+
+    s = Dict([(Tuple(var_index(x)), x) for x in s])
```
`s` changes its type here, from vector to dict, which makes the code type unstable. Just rename one. And while at it, rename `s_var_name`... ? How about this then
```suggestion
function gaussian_ring(n::Int, s::VarName="s", base_ring::Field=QQ)

    S, v = polynomial_ring(base_ring, reduce(vcat, [[s_var_name*string([i,j]) for j in i:n] for i in 1:n]))

    d = Dict([(Tuple(var_index(x)), x) for x in v)])
```
and then use `d` in the next two lines, too

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+    E = [(src(e), dst(e)) for e in edges(G)]
+
+    if length(E) == 0
+        print("Directed Gaussian graphical model on the empty graph")
```
Please check all your `print` statements as to whether they take `io` into account

taboege replied:
Done

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,412 @@
+
+
+@doc raw"""
+    parameterization(M::GraphicalModel{Graph{Directed}, GaussianRing})
```
while both "parametrization" and "parameterization" are valid spellings, elsewhere we already use `parametrization` so I would suggest following that (CC @JohnAAbbott and @mjrodgers just in case our native speakers have a different view on this)
```suggestion
    parameterization(M::GraphicalModel{Graph{Directed}, GaussianRing})
```

taboege replied:
Done.

fingolfin commented on experimental/AlgebraicStatistics/docs/src/ci.md:
```diff
@@ -0,0 +1,32 @@
+CurrentModule = Oscar
+DocTestSetup = quote
+  using Oscar
+end
```
You cannot know this (because it was only *just now* changed), but this should now be
```suggestion
DocTestSetup = Oscar.doctestsetup()
```
also in the other .md files.

taboege replied:
Done

fingolfin commented on experimental/AlgebraicStatistics/src/GraphicalModels.jl:
```diff
@@ -0,0 +1,129 @@
+    str = string(x)
+    index = "["*split(str, "[")[2]
+
+    eval(Meta.parse(index))
```
Oh wow, that's the big guns! :-)

I am hopefully you don't need a full Julia parser here! At what kinds of strings are we looking here? If you tell us what typical `index` values are, I am sure we can suggest a less extreme way to parse it :-)

bkholler replied:
Haha! 

I basically expect my string to look like "x[i, j, k, l]" where i,j,k,l are just integers. The main purpose of this though was just to take my variable x[i,j,k,l] which is a generator of a multivariate polynomial ring and return [i,j,k,l] as Vector{Int}. 

I basically wrote this to mimic the function "last baseName" in Macaulay2 but I would love a better way to do it!

taboege replied:
@bkholler already explained the reasoning behind this: we would like to be able to extract the object which indexes a given variable (usually a `Vector`) from that variable object, e.g. the `[1,2]` as an actual `Int[]` from a generator `s[1,2]` in a polynomial ring.

In the meantime, I have changed all the places which use the function quote above to generate a list of indices. From that list of indices, we generate the variable names for the generators of `polynomial_ring` as well as the dictionary. No parsing needed anymore.

fingolfin commented on experimental/AlgebraicStatistics/src/GraphicalModels.jl:
```diff
@@ -0,0 +1,129 @@
+end
+
+
+function adjoint(A::Generic.MatSpaceElem)
```
In Julia, `adjoint(A)` is defined as `conjugate transposition`, I think it would be dangerous to repurpose it here. 

I think what you write here is actually the ["adjugate matrix"](https://en.wikipedia.org/wiki/Adjugate_matrix), isn't it?

fingolfin commented on experimental/AlgebraicStatistics/src/PhylogeneticAuxiliary.jl:
```diff
@@ -0,0 +1,89 @@
+##########################
+#### GROUP OPERATIONS ####
+##########################
+# This function implements the group operation on Z2 x Z2 and is subsequently 
```
Have you considered using `abelian_group(2,2)` for this?

Of course maybe it is better for you to work like you do right now, I am just wondering.

marinagarrote replied:
We did not, we didn't know, but it can be useful. Thanks!

fingolfin commented on experimental/AlgebraicStatistics/src/PhylogeneticAuxiliary.jl:
```diff
@@ -0,0 +1,89 @@
+    in_node = inneighbors(graph,l)[1]
+    lvs_cherr = outneighbors(graph, inneighbors(graph,l)[1])
+    if issubset(lvs_cherr, lvs) == 2
+      cherry = append!(cherry, [[Edge(in_node, lvs_cherr[1]), Edge(in_node, lvs_cherr[2])]])
```
```suggestion
      append!(cherry, [[Edge(in_node, lvs_cherr[1]), Edge(in_node, lvs_cherr[2])]])
```

taboege replied:
Done

fingolfin commented on experimental/AlgebraicStatistics/src/PhylogeneticAuxiliary.jl:
```diff
@@ -0,0 +1,89 @@
+    if issubset(lvs_cherr, lvs) == 2
+      cherry = append!(cherry, [[Edge(in_node, lvs_cherr[1]), Edge(in_node, lvs_cherr[2])]])
+    end
+  end
```
Should this function `return cherry`?

fingolfin commented on experimental/AlgebraicStatistics/src/PhylogeneticAuxiliary.jl:
```diff
@@ -0,0 +1,89 @@
+
+function cherries(graph::Graph)
+  lvs = leaves(graph)
+  cherry = []
```
`cherry` is now a `Vector{Any}`, can you make the type more precise?

taboege replied:
Yes, done.

fingolfin reviewed (commented):
Here is a bunch of comments. Please don't be intimidated -- overall your code looks nice, and almost all of my comments are for minor things. Feel free to question any and all of them, they are intended as helpful hints, not to block you or this PR.

fingolfin commented:
Is this meant to be reviewed more carefully and then merged? Then the first necessary step would be to make CI tests pass.

Or is this still work in progress? Then perhaps use the "Convert to draft" link :-)

taboege commented:
Many thanks for your comments, Max! This is indeed intended to be merged. We'll try to go through the suggestions before the weekend.

taboege committed:
Run bibtool

taboege committed:
Use `io` in Base.show() methods

taboege committed:
Use doctestsetup

As in 7bf8d95ba2.

taboege committed:
Fix cherries()

Address some of the review comments from #3812 about the cherries()
function and correct the non-sensical predicate `issubset(A,B) == 2`.
It is the *intersection* of A and B that should have size 2.

taboege committed:
Resolve errors in Documentation CI tests

taboege committed:
Make tests and doctests pass

- Add `export`s to make tests pass.
- Also get rid of Meta.parse in var_index as suggested by @fingolfin.
- Update expected outputs in doctests.

taboege committed:
Export structs

Them being unexported makes doctests fail in which type names are printed.

bkholler committed:
removed DiscreteGraphicalModels.jl

taboege committed:
Always `collect` arguments to `subsets`

`subsets` currently works with `Vector` and `Set`, but nothing iterable,
in particular not with ranges.

taboege committed:
Fix doctest

Commit:
Abelian group in group based models

Commit:
Merge branch 'g/groups_GBmod' into bgh/AlgebraicStatistics

taboege committed:
Rename adjoint -> adjugate

taboege committed:
Fix some doctests in AlgebraicStatistics

Commit:
changed tests for groups in GB models

bkholler commented:
@fingolfin Sorry to bother you again but it seems that the only tests were now failling are due to our doctests not working since they were made in groups. Is it possible to have these kind of grouped doctests in OSCAR or should we change them so that each doctest has all of the objects it needs defined and is not part of a group?

fingolfin commented:
OSCAR does nothing to stop you from using Documenter.jl's doctest grouping features.

However, I would recommend against using that to group doctests across different docstrings. Because a user will only see one docstring and the doctest(s) in it, and have no way to figure out which other docstring may contain the code needed to get the doctest in the first docstring working. So then your examples become useless.

Commit:
changed order entries group

Commit:
Change structure of equivalent classes

Commit:
Documentation new equivalent classes

bkholler committed:
changed doctests to no longer depend on grouping

bkholler committed:
fixed small errors in output for doctests in GaussianGraphicalModels.jl and Markov.jl

bkholler committed:
fixed failing doctests in CI.jl and Markov.jl relating to printing ci stmts

Commit:
Specialized Fourier transform for new structure equivalent classes

Commit:
sum equivalent classes for new structure equivalent classes

Commit:
Merge branch 'bgh/AlgebraicStatistics' of github.com:bkholler/Oscar.jl into bgh/AlgebraicStatistics

Commit:
Fix failing doctests sum_queivalent_classes and group_of_model

Commit:
Fix character in failing doctests sum_queivalent_classes and group_of_model

Commit:
Fixed phylogenetics.md

Commit:
Fixed type @doc specialized_fourier_transform

Commit:
Fixed phylogenetics.md specialized_fourier_transform

fieker commented on experimental/AlgebraicStatistics/src/GraphicalModels.jl:
```diff
@@ -0,0 +1,119 @@
+###################################################################################
+
+
+function cofactor(A)
```
maybe use `pseudo_inv` to compute this? Should be much faster than all the single det calls

taboege replied:
Thanks, will do.

lkastner commented:
Since @antonydellavecchia is gone until the beginning of July, would you want to wait for him or should we go ahead? @bkholler

I think there is not much to do before this can have a first merge, also it is going to experimental, so modification will still be allowed. Main points would be the README.md and the weird empty lines. I'll do a detailed review if you want me to.

fingolfin commented on experimental/AlgebraicStatistics/src/CI.jl:
```diff
@@ -0,0 +1,225 @@
+
+function Base.show(io::IO, stmt::CIStmt)
+  fmt(K) = length(K) == 1 ? string(K[1]) : "{" * join([string(x) for x in K], ", ") * "}"
+  print(io, "[$(fmt(stmt.I)) ⫫ $(fmt(stmt.J)) | $(fmt(stmt.K))]")
```
Unconditional use of unicode here goes against our conventions, see https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/#Unicode and https://docs.oscar-system.org/dev/DeveloperDocumentation/printing_details/#Unicode-printing for details

taboege replied:
Thanks for the reminder. I changed that to conditional.

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
```diff
@@ -0,0 +1,454 @@
+
+    varindices = [Tuple([i,j]) for i in 1:n for j in i:n]
+    varnames = ["$(s_var_name)[$(i), $(j)]" for (i,j) in varindices]
+    S, s = polynomial_ring(base_ring, varnames)
```
Probably should add `cached=false` to most/all `polynomial_ring` invocations, see also https://docs.oscar-system.org/dev/DeveloperDocumentation/caching/#Rules-for-implementations (and a bunch of recent PRs which add missing `cached=false` args in many places)

taboege replied:
Done.

fingolfin commented:
There are still a bunch of code comments I made that are not addressed -- one may have to click on "Load more..." under "84 hidden items") multiple times to see them all. Even then it may be difficult to spot them among the many comments which already have been handled. Perhaps @bkholler can use "Resolve conversation" button on all comments he considers resolved?

fingolfin commented on experimental/AlgebraicStatistics/src/GaussianGraphicalModels.jl:
````diff
@@ -0,0 +1,454 @@
+s[1, 1], s[1, 2], s[1, 3], s[2, 2], s[2, 3], s[3, 3]
+```
+"""
+function gaussian_ring(n::Int, s_var_name::String="s", base_ring::Field=QQ)
````
You are hiding the function `base_ring` here -- I'd recommend to call this just `R`.

And I still would suggest to change `::String` to `::VarName`

taboege replied:
Both done.

fingolfin commented:
(I actually now marked a bunch of the old comments as resolved where it was clear that they were. Still leaves a couple unaddressed ones)

taboege committed:
Reformat code

Removed empty lines and adapted indentation size to conform to .editorconfig.

taboege committed:
Make printing of CIStmt unicode-free by default

taboege committed:
Use pseudo_inv() to implement cofactor()

Should be more efficient than computing all of those cofactors separately.

taboege committed:
Do not shadow base_ring

Also force the coefficient ring of MarkovRing and GaussianRing to be
of type `Field` everywhere.

taboege committed:
Type variable names as VarName

taboege committed:
Make more Unicode printing conditional

taboege committed:
Remove useless mentions of "Oscar" from documentation

taboege committed:
Strengthen types in MarkovRing and GaussianRing

Also make the interfaces to their `gens` similar: both are now accessed
through a Dict mapping indices to generators of the polynomial ring.

taboege committed:
Use "parametrization" spelling

Instead of "parameterization".

taboege committed:
Do not cache polynomial rings by default

taboege committed:
AlgebraicStatistics: add README

taboege committed:
Fix Unicode-y doctests

- `CIStmt` printing had inverted test for `Oscar.is_unicode_allowed()`.
- `ci_statements()` and `ci_stmt()` didn't accept `VarName`.

taboege commented:
I have addressed all of the comments and implemented the suggested changes, although it doesn't seem like I can mark any of the review threads as resolved.

@fingolfin From your tips, I take it that adding type information to `struct`s and method arguments in Julia is beneficial. In the last batch of commits, I have added some more than the ones you suggested, including types such as `Vector{<:VarName}` or `Pair{<:VarName, <:AbstractArray}` where type parameters are constrained. I hope that this kind of nested type annotation is also beneficial.

lkastner commented on experimental/AlgebraicStatistics/src/Markov.jl:
```diff
@@ -0,0 +1,331 @@
+          push!(eqs, p11*p22 - p12*p21)
+        end
+      end
+    end
```
```suggestion
    for z in state_space(R, stmt.K), x in subsets(collect(QI), 2), y in subsets(collect(QJ), 2)
        p11 = marginal(R, IJK, [x[1], y[1], z...]);
        p12 = marginal(R, IJK, [x[1], y[2], z...]);
        p21 = marginal(R, IJK, [x[2], y[1], z...]);
        p22 = marginal(R, IJK, [x[2], y[2], z...]);
        push!(eqs, p11*p22 - p12*p21)
    end
```
The loops seem not to depend on each other, right?

lkastner commented on experimental/AlgebraicStatistics/src/PhylogeneticAuxiliary.jl:
```diff
@@ -0,0 +1,71 @@
+#### AUXILIARY FUNCTIONS FOR GRAPHS ####
+########################################
+
+function interior_nodes(graph::Graph)
+  big_graph = Polymake.graph.Graph(ADJACENCY = pm_object(graph))
+  degrees = big_graph.NODE_DEGREES
+  return findall(x -> x > 1, degrees)
+end
+
+function leaves(graph::Graph)
+  big_graph = Polymake.graph.Graph(ADJACENCY = pm_object(graph))
+  degrees = big_graph.NODE_DEGREES
+  return findall(x -> x == 1, degrees)
+end
+
+function vertex_descendants(v::Int, gr::Graph, desc::Vector{Any})
+  lvs = leaves(gr)
+  outn = outneighbors(gr, v)
+
+  if v in lvs
+    return [v]
+  end
+
+  innodes = setdiff(outn, lvs)
+  d = unique(append!(desc, intersect(outn, lvs)))
+  
+  if length(innodes) > 0
+    for i in innodes
+      d = vertex_descendants(i, gr, d)
+    end
+    return d 
+  end
+
+  return d
+end
+
+function root(graph::Graph)
+  n_parents = [length(inneighbors(graph, v)) for v in 1:n_vertices(graph)]
+  return findall(x -> x == 0, n_parents)[1]
+end
```
Is it important whether these are directed or undirected? Because when you convert you always get something undirected:
```
julia> g = Graph{Directed}(5)
Directed graph with 5 nodes and no edges

julia> add_edge!(g,1,2)
true

julia> pmg = Polymake.graph.Graph(ADJACENCY=Oscar.pm_object(g))
type: Graph<Undirected>

ADJACENCY
	{1}
	{0}
	{}
	{}
	{}
```
You could instead forward the template argument:
```
julia> pmg = Polymake.graph.Graph{Directed}(ADJACENCY=Oscar.pm_object(g))
type: Graph<Directed>

ADJACENCY
	{1}
	{}
	{}
	{}
	{}
```

lkastner commented on experimental/AlgebraicStatistics/src/CI.jl:
```diff
@@ -0,0 +1,229 @@
+end
+
+@doc raw"""
+  ci_stmt(I::Vector{<:VarName}, J::Vector{<:VarName}, K::Vector{<:VarName}; symmetric=true, semigraphoid=true)
```
```suggestion
@doc raw"""
    ci_stmt(I::Vector{<:VarName}, J::Vector{<:VarName}, K::Vector{<:VarName}; symmetric=true, semigraphoid=true)
```
The signature needs to be indented by exactly four spaces, otherwise it does not show correctly in the documentation. Try
```
build_doc(doctest=false)
```
to see the docs locally.

lkastner replied:
This is the only occurrence of wrongly indented signature I found btw when building the docs locally, so unless you want to have a look at other parts of the documentation there is probably no need for you to build it locally.

lkastner reviewed (approved):
I had some minor comments, but overall I think this can be merged to experimental now and then refined later.

We might remove the new docs for `Base.:(==)`, `Base.hash` and `gens`, since they are fairly small and the methods are self-explanatory. I think we proceeded in the same way for other parts of Oscar.

taboege committed:
Fix documentation and combine independent loops

taboege commented:
Thank you, Lars! I think the bit about the docs indentation is important even for a first merge, so I fixed that and pushed a new commit.

And while at it, I also combined the three independent loops as you suggested. I can't speak about the phylogenetics code. That is @marinagarrote's area. Other than that, @bkholler is on vacation but our consensus was to get a first merge as soon as possible and then continue straightening things out and implementing more features.

fingolfin reviewed (approved):
Thank you. I cannot evaluate the mathematics in here. But all my code comments were addressed quickly and competently, and the code now looks to be in quite good shape.
